src/groups.py

```diff
@@ -0,0 +1,380 @@
+            if key == service_name:
+                yaml_string = yaml.dump(data)
+                relation.data[self.charm.app].update(
+                    {"user-group-configuration": yaml_string}
+                )
+
+    def _validate(self):  # noqa: C901
+        """Validate user-group-configuration file values.
+
+        Raises:
+            ValueError: In case the file cannot be parsed.
+                        In case the file cannot be converted to a dictionary.
+                        In case there are no related services.
+                        In case relation key is missing.
+                        In case service name has no corresponding relation.
+                        In case user, group or membership values are missing.
+
+        """
+        # Validate data can be loaded from the file.
+        try:
+            data = yaml.safe_load(
+                self.charm.config["user-group-configuration"]
+            )
+        except yaml.YAMLError as e:
+            raise ValueError(
+                "The configuration file is improperly formatted, unable to parse."
+            ) from e
+
+        # Validate resulting data is a dictionary.
+        if not isinstance(data, dict):
+            raise ValueError("The configuration file is improperly formatted.")
+
+        # Validate there are policy relations.
+        policy_relations = self.charm.model.relations.get("policy")
+        if not policy_relations:
+            raise ValueError(
+                "There are no relations with which to apply this file."
+            )
+
+        service_names = []
+        for relation in policy_relations:
+            service_name = relation.data[self.charm.app].get("service_name")
+            service_names.append(service_name)
+
+        for key in data.keys():
+            # Validate the file has a service name.
+            if any(keyword in key for keyword in EXPECTED_KEYS):
+                raise ValueError(
+                    "User management configuration file must have service keys."
+                )
+
+            # Validate the file contains only services that exist.
+            if key not in service_names:
+                raise ValueError(f"{key} does not match a related service.")
+
+             # Validate that there are `user`, `group` and `membership` keys.
+            for expected_key in EXPECTED_KEYS:
+                if expected_key not in data[key]:
+                    raise ValueError(
+                        f"Missing '{expected_key}' values in the configuration file."
+                    )
+
+
+def create_payload(member_type, member):
+    """Create a payload for a new user in the Ranger API.
+
+    Args:
+        member_type: The type of Ranger member (group, user or membership).
+        member: The fields for creating the member.
+
+    Returns:
+        user_payload: User payload data for creating a new user.
+    """
+    if member_type == "user":
```

Comment:
Inconsistent spacing between `if` statements. By the way, if the Python version is new enough (3.10<=), you can use a `match` statement here.

Reply:
thank you, I have updated the if spacing but left with if statements as the tox python used for testing the code is not new enough